main_test.go

```diff
@@ -10,16 +19,392 @@ func TestNormalizeInstanceName(t *testing.T) {
 	}
 }
 
-func TestMakeName(t *testing.T) {
-	expected := "bs-name"
-	actual := makeName("bs", "name")
-	if actual != expected {
-		t.Fatalf("%s is not equal %s", actual, expected)
+type handleServiceUpdatesTestCase struct {
+	Update                       *registry.ServiceUpdate
+	CreateNEGCalls               []spy.CreateNEGCall
+	CreateHCCalls                []spy.CreateHCCall
+	CreateBSCalls                []spy.CreateBSCall
+	UpdateURLMapCalls            []spy.UpdateURLMapCall
+	AttachOrDetachEndpointsCalls []spy.AttachOrDetachEndpointsCall
+}
+
+type handleServiceUpdatesTestSuite []handleServiceUpdatesTestCase
+
+func testHandleServiceUpdates(
+	t *testing.T,
+	urlMapName string,
+	tagConfig *tagConfiguration,
+	testSuite handleServiceUpdatesTestSuite,
+) {
+	timeToProcessUpdate := 250 * time.Millisecond
+	cloudClient := &spy.SpyCloudClient{}
+
+	updates := make(chan *registry.ServiceUpdate)
+	var wg sync.WaitGroup
+	done := make(chan struct{})
+
+	wg.Add(1)
+	go handleServiceUpdates(urlMapName, cloudClient, tagConfig, updates, &wg, done)
+
+	for _, testCase := range testSuite {
+		updates <- testCase.Update
+
+		time.Sleep(timeToProcessUpdate)
+
+		if diff := deep.Equal(cloudClient.PopCreateNEGCalls(), testCase.CreateNEGCalls); diff != nil {
+			t.Fatalf("%+v\n", diff)
+		}
+
+		if diff := deep.Equal(cloudClient.PopCreateHCCalls(), testCase.CreateHCCalls); diff != nil {
+			t.Fatalf("%+v\n", diff)
+		}
+
+		if diff := deep.Equal(cloudClient.PopCreateBSCalls(), testCase.CreateBSCalls); diff != nil {
+			t.Fatalf("%+v\n", diff)
+		}
+
+		if diff := deep.Equal(cloudClient.PopUpdateURLMapCalls(), testCase.UpdateURLMapCalls); diff != nil {
+			t.Fatalf("%+v\n", diff)
+		}
+
+		if diff := deep.Equal(cloudClient.PopAttachOrDetachEndpointsCalls(), testCase.AttachOrDetachEndpointsCalls); diff != nil {
+			t.Fatalf("%+v\n", diff)
+		}
```

Comment:
Good candidate for a switch statement.

Reply:
Though structure of the underlying client makes me wonder, if you can re-arrange it to get away with a single deep equal comparisson.